flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -0,0 +1,301 @@
+			FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+			ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+			CreditBasedInputBuffersUsageGauge inputBuffersUsageGauge = new CreditBasedInputBuffersUsageGauge(
+				floatingBuffersUsageGauge,
+				exclusiveBuffersUsageGauge,
+				inputGates);
+
+			assertEquals(0.0, floatingBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, inputBuffersUsageGauge.getValue(), 0.0);
+
+			// drain gate1's exclusive buffers
+			drainBuffer(buffersPerChannel, remoteInputChannel1, closeableRegistry);
+			drainBuffer(buffersPerChannel, remoteInputChannel2, closeableRegistry);
+
+			int totalBuffers = extraNetworkBuffersPerGate * numberOfInputGates + buffersPerChannel * numberOfRemoteChannelsTotal;
+
+			remoteInputChannel1.requestSubpartition(0);
+
+			int backlog = 3;
+			int requestFloatingTotal = buffersPerChannel + backlog;
+
+			remoteInputChannel1.onSenderBacklog(backlog);
+
+			assertEquals(requestFloatingTotal, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
+
+			drainBuffer(requestFloatingTotal, remoteInputChannel1, closeableRegistry);
+
+			assertEquals(0, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
+			assertEquals((double) (buffersPerChannel * numberOfRemoteChannelsGate1 + requestFloatingTotal) / totalBuffers,
+				inputBuffersUsageGauge.getValue(), 0.0001);
+		} finally {
+			inputGate1.close();
+			inputGate2.close();
+			closeableRegistry.close();
+			network.close();
+		}
+	}
+
+	private void drainBuffer(int boundary, RemoteInputChannel channel, CloseableRegistry closeableRegistry) throws IOException {
+		for (int i = 0; i < boundary; i++) {
+			Buffer buffer = channel.requestBuffer();
+			if (buffer != null) {
+				closeableRegistry.registerCloseable(buffer::recycleBuffer);
+			} else {
+				break;
+			}
+		}
+	}
+
+	private RemoteInputChannel buildRemoteChannel(
+		int channelIndex,
+		ResultPartitionID partitionID,
+		SingleInputGate inputGate,
+		NettyShuffleEnvironment network) {
+		return new InputChannelBuilder()
+			.setChannelIndex(channelIndex)
+			.setPartitionId(partitionID)
+			.setupFromNettyShuffleEnvironment(network)
+			.setConnectionManager(new TestingConnectionManager())
+			.buildRemoteAndSetToGate(inputGate);
+	}
+
+	private LocalInputChannel buildLocalChannel(
+		int channelIndex,
+		ResultPartitionID partitionID,
+		SingleInputGate inputGate,
+		NettyShuffleEnvironment network) {
+		return new InputChannelBuilder()
+			.setChannelIndex(channelIndex)
+			.setPartitionId(partitionID)
+			.setupFromNettyShuffleEnvironment(network)
+			.setConnectionManager(new TestingConnectionManager())
+			.buildLocalAndSetToGate(inputGate);
+	}
+
+	private Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> buildInputGate(
+		NettyShuffleEnvironment network,
+		int remoteChannelNum,
```

Comment:
remoteChannelNum -> numberOfRemoteChannels, also for the following local channel